src/test/ui/issues/issue-7813.stderr

```diff
@@ -1,4 +1,4 @@
-error[E0282]: type annotations needed
+error[E0282]: type annotations needed in `&[_; 0]`
```

Comment:
Why `_` here...?

Reply:
We don't always have access to the `TyInfer` name. I think it is because `[]` is an intrinsic that doesn't _have_ a name anywhere. Due to this, I would expect this output only from arrays, slices and tuples.

Reply:
So since it's not always reliably possible to not use `_` I would consider always using `_` instead.

Reply:
When we only have `_`, we don't mention any type argument name anywhere:

```
error[E0282]: type annotations needed in `[_; 0]`
  --> $DIR/cannot_infer_local_or_array.rs:2:13
   |
LL |     let x = [];
   |         -   ^^ cannot infer type
   |         |
   |         consider giving `x` a type
```

We could change the above to be

```
error[E0282]: type annotations needed in `[_; 0]`
  --> $DIR/cannot_infer_local_or_array.rs:2:13
   |
LL |     let x = [];
   |         -   ^^ cannot infer type `_` in `[_; 0]`
   |         |
   |         consider giving `x` a type
```

but the underlying situation is still there. If we only use `_` we could have the following situation:

```
error[E0282]: type annotations needed in `std::result::Result<i32, E>`
 --> file7.rs:3:13
  |
3 |     let b = foo();
  |         -   ^^^ cannot infer type for `_` in `std::result::Result<_, _>`
  |         |
  |         consider giving `b` a type
```

instead of

```
error[E0282]: type annotations needed in `std::result::Result<A, B>`
 --> file7.rs:6:13
  |
6 |     let b = foo();
  |         -   ^^^ cannot infer type for `A` in `std::result::Result<A, B>`
  |         |
  |         consider giving `b` a type

```

which in my eyes is more confusing. A case that illustrates this nicely is the following, which is now

```
error[E0282]: type annotations needed
  --> $DIR/issue-25368.rs:11:17
   |
LL |     let (tx, rx) = channel();
   |         -------- consider giving the pattern a type
...
LL |         tx.send(Foo{ foo: PhantomData });
   |                 ^^^ cannot infer type for `_`
```

and after this PR would now be

```
error[E0282]: type annotations needed in `(std::sync::mpsc::Sender<Foo<T>>, std::sync::mpsc::Receiver<Foo<T>>)`
  --> $DIR/issue-25368.rs:11:17
   |
LL |     let (tx, rx) = channel();
   |         -------- consider giving the pattern a type
...
LL |         tx.send(Foo{ foo: PhantomData });
   |                 ^^^ cannot infer type for `T` in `(std::sync::mpsc::Sender<Foo<T>>, std::sync::mpsc::Receiver<Foo<T>>)`
```

making it slightly more clear that `T` needs to be replaced in both sides of the tuple.

Reply:
Fair enough.